src/cpp/server_core/ServerDatabase.cpp

```diff
@@ -0,0 +1,204 @@
+/*
+ * ServerDatabase.cpp
+ *
+ * Copyright (C) 2020 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <server_core/ServerDatabase.hpp>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/thread.hpp>
+
+#include <core/Log.hpp>
+#include <core/Settings.hpp>
+#include <core/system/Environment.hpp>
+#include <core/system/System.hpp>
+#include <shared_core/Error.hpp>
+
+namespace rstudio {
+namespace server_core {
+namespace database {
+
+using namespace core;
+using namespace core::database;
+
+namespace {
+
+// settings constants
+constexpr const char* kDatabaseProvider = "provider";
+constexpr const char* kDatabaseProviderSqlite = "sqlite";
+constexpr const char* kDatabaseProviderPostgresql = "postgresql";
+constexpr const char* kSqliteDatabaseDirectory = "directory";
+constexpr const char* kDefaultSqliteDatabaseDirectory = "/var/run/rstudio-server";
+constexpr const char* kDatabaseHost = "host";
+constexpr const char* kDefaultDatabaseHost = "localhost";
+constexpr const char* kDatabasePort = "port";
+constexpr const char* kDefaultPostgresqlDatabasePort = "5432";
+constexpr const char* kDatabaseUser = "user";
+constexpr const char* kDefaultPostgresqlDatabaseUser = "postgres";
+constexpr const char* kDatabasePassword = "password";
+constexpr const char* kPostgresqlDatabaseConnectionTimeoutSeconds = "connnection-timeout-seconds";
+constexpr const int   kDefaultPostgresqlDatabaseConnectionTimeoutSeconds = 10;
+
+// environment variables
+constexpr const char* kServerTmpDirEnvVar = "RS_SERVER_TMP_DIR";
+constexpr const char* kDatabaseMigrationsPathEnvVar = "RS_DB_MIGRATIONS_PATH";
+
+//misc constants
+constexpr const size_t kDefaultConnectionPoolSize = 4;
+
+boost::shared_ptr<ConnectionPool> s_connectionPool;
```

Comment:
Does `rserver` still exit gracefully with this static variable?

Reply:
Yes, any thread that could hold a reference to this will be stopped before it exits.